File: lzw/lzw.py
import os
import pickle

DICTIONARY_SIZE = 256

def compress(input):
    global DICTIONARY_SIZE
    dictionary = {}
    result = []
    temp = ""

    for i in range(0, DICTIONARY_SIZE):
        dictionary[str(chr(i))] = i

    for c in input:
        temp2 = temp + str(chr(c))
        if temp2 in dictionary.keys():
            temp = temp2
        else:
            result.append(dictionary[temp])
            dictionary[temp2] = DICTIONARY_SIZE
            DICTIONARY_SIZE += 1
            temp = "" + str(chr(c))

    if temp != "":
        result.append(dictionary[temp])

    return result

def decompress(input):
    global DICTIONARY_SIZE
    dictionary = {}
    result = []

    for i in range(0, DICTIONARY_SIZE):
        dictionary[i] = str(chr(i))

    previous = chr(input[0])
    input = input[1:]
    result.append(previous)

    for bit in input:
        aux = ""
        if bit in dictionary.keys():
            aux = dictionary[bit]
        else:
            aux = previous + previous[0]

        result.append(aux)
        dictionary[DICTIONARY_SIZE] = previous + aux[0]
        DICTIONARY_SIZE += 1
        previous = aux

    return result

if __name__ == "__main__":
    ABSOLUTE_PATH = os.getcwd()

    # Access arguments from args.py
    from args import arguments

    if arguments.action == 'compress':
        input_data = open(ABSOLUTE_PATH + "//" + arguments.input, "rb").read()
        output_file = open(ABSOLUTE_PATH + "//" + arguments.output, "wb")

        compressed_data = compress(input_data)
        pickle.dump(compressed_data, output_file)
    else:
        input_data = pickle.load(open(ABSOLUTE_PATH + "//" + arguments.input, "rb"))
        output_file = open(ABSOLUTE_PATH + "//" + arguments.output, "w", encoding='utf-8') 

        uncompressed_data = decompress(input_data)
        for line in uncompressed_data:
            output_file.write(line)

        output_file.close()

File: lzw/lzw.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>

const int DICTIONARY_SIZE = 256;

std::vector<int> compress(const std::string& input) {
    std::unordered_map<std::string, int> dictionary;
    std::vector<int> result;
    std::string temp;

    // Initialize dictionary with single character entries
    for (int i = 0; i < DICTIONARY_SIZE; ++i) {
        dictionary[std::string(1, char(i))] = i;
    }

    int dictSize = DICTIONARY_SIZE;

    for (char c : input) {
        std::string temp2 = temp + c;
        if (dictionary.find(temp2) != dictionary.end()) {
            temp = temp2;
        } else {
            result.push_back(dictionary[temp]);
            dictionary[temp2] = dictSize++;
            temp = std::string(1, c);
        }
    }

    if (!temp.empty()) {
        result.push_back(dictionary[temp]);
    }

    return result;
}

std::string decompress(const std::vector<int>& input) {
    std::unordered_map<int, std::string> dictionary;
    std::string result;

    // Initialize dictionary with single character entries
    for (int i = 0; i < DICTIONARY_SIZE; ++i) {
        dictionary[i] = std::string(1, char(i));
    }

    int dictSize = DICTIONARY_SIZE;

    std::string temp = dictionary[input[0]];
    result += temp;

    for (size_t i = 1; i < input.size(); ++i) {
        int current = input[i];
        std::string entry;

        if (dictionary.find(current) != dictionary.end()) {
            entry = dictionary[current];
        } else {
            entry = temp + temp[0];
        }

        result += entry;

        // Add entry to dictionary
        dictionary[dictSize++] = temp + entry[0];

        temp = entry;
    }

    return result;
}

int main() {
    std::string input_data;

    // Prompt user for input
    std::cout << "Enter the input string: ";
    std::getline(std::cin, input_data);

    // Compress the input data
    std::vector<int> compressed_data = compress(input_data);

    // Decompress the compressed data
    std::string decompressed_data = decompress(compressed_data);

    // Output results
    std::cout << "Original: " << input_data << std::endl;
    std::cout << "Compressed: ";
    for (int code : compressed_data) {
        std::cout << code << " ";
    }
    std::cout << std::endl;
    std::cout << "Decompressed: " << decompressed_data << std::endl;

    return 0;
}
